lib/swim.ml

```diff
@@ -0,0 +1,127 @@
+open Base.Hashtbl
+
+type message =
+  | Ping
+  | Acknowledge
+  | PingRequest of Peer.address
+[@@deriving bin_io]
+
+type config = {
+  protocol_period : int;
+  round_trip_time : int;
+  timeout : int;
+}
+
+type t = {
+  config : config;
+  (* TODO: I am totally not sure about this type *)
+  acknowledges : (int, unit Lwt_condition.t) Poly.t;
+  mutable peers : Peer.t list;
+  (* TODO: I am not sure this is useful, it's just a counter to know what is the current round
+     I would like to send it with the message payload, what do you think? *)
+  mutable sequence_number : int;
+}
+
+(* Helper to increase the round *)
+let next_seq_no t =
+  let sequence_number = t.sequence_number in
+  t.sequence_number <- t.sequence_number + 1;
+  sequence_number
+
+let send_ping_to client (peer_address : Peer.address) =
```

Comment:
Probably can abstract into a helper function like

```ocaml
let send_msg msg client peer_address =
  let msg = Util.Encoding.pack bin_writer_message msg in
  Client.send_to client msg peer_address
```

So that we can do e.g.

```ocaml
let send_ping_to = send_msg Ping
```

Reply:
This way we can also write any header-injection logic only once :+1:

Reply:
Yup, nice idea! Done!